rfcs/2022-11-04-automatic-component-validation.md

```diff
@@ -0,0 +1,611 @@
+
+In the most basic test case, we can trivially provide simple log events that will be encoded and
+processed by the source without issue. However, in order to exercise the failure path, we need to
+consider what causes the source to fail to process an event.
+
+For the `http_server` source, the only real failure path is a payload that cannot be decoded. From
+the definitions we've seen above, the only control that we really have is over the contents of the
+event itself. As components describe their own configuration, and the configuration of their
+external resource, in a fixed way, this means that we could, for example, only specify the external
+resource to create with a fixed codec, such as JSON. The HTTP resource code is configured with this
+information, and will dutifully take an `Event` and encode it to JSON before sending it along.
+
+To get around this, `TestEvent` is introduced and provides two modes: passthrough and modified. The
+passthrough mode is exactly what it sounds like: the event it holds is used as-is. When faced with a
+situation such as our example, where modifying the event itself cannot be used to induce a failure,
+we can utilize the modified mode to do so.
+
+The modified mode also wraps an `Event`, but it is used as a signal to an external resource that the
+external resource should modify the event in some context-specific way in order to go against the
+happy path. As you might surmise, this is generally only useful for sources, because transforms and
+sinks will always be given an `Event`, and so the event itself must be tailored to induce failure in
+thos cases.
+
+Going back to our example, we can craft a "modified" event which will instruct the HTTP client
+external resource to encode the event using an encoding that _isn't_ the one it was configured with.
+Codifying this in the definition of the test cases for the `http_server` source would look something
+like this:
+
+```rust
+impl ValidatableComponent for SimpleHttpConfig {
+    ...
+
+    fn test_cases(&self) -> Vec<TestCase> {
+        vec![
+          // This is a happy path test, and each event will be encoded according to the external resource definition.
+          TestCase::success(vec![
+            LogEvent::from_str_legacy("simple message 1"),
+            LogEvent::from_str_legacy("simple message 2"),
+            LogEvent::from_str_legacy("simple message 3"),
+          ]),
+
+          // This is a failure path test, where we've marked one event such that the external resource should modify
+          // it, which in this case will mean it gets encoded using a codec other than the one specified in the
+          // external resource definition.
+          TestCase::partial_failure(vec![
+            LogEvent::from_str_legacy("good message 1").into(),
+            TestEvent::modified(LogEvent::from_str_legacy("bad message 1")),
+            LogEvent::from_str_legacy("good message 2").into(),
+          ]),
+        ]
+    }
```

Comment:
Yeah, that was a thought I had as well. Test scenarios sound like data more than running resources.

Reply:
I changed this in cf266d9c3.

Now every component has its test cases through via YAML file at a well-known path, based on the component type and component name: `tests/validation/components/<component type>/<component name>.yaml`.

You can look at `tests/validation/components/sources/http_server.yaml` for an example of what this looks like. I'm taking an approach that uses some custom types to massage the (de)serialization so that we don't have to write out all the boilerplate that would be necessary to deserialize directly to `Event`.

It still needs to be expanded to cover all metric types, though.